File: cses/08_string_algorithms/1149_distinct-subsequences/solution.py
import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        sys.stdout.write("0\n")
        return
    s = data[0]

    MOD = 1_000_000_007
    dp = 1               # dp: số subsequence phân biệt tính cả chuỗi rỗng
    last = [0] * 26      # last[c] = dp[j-1] của lần xuất hiện gần nhất của c
    seen = [False] * 26

    for byte in s:
        c = byte - 97    # 'a' == 97
        # Gấp đôi: mỗi chuỗi cũ có thể giữ nguyên hoặc gắn thêm ký tự hiện tại.
        new = (2 * dp) % MOD
        # Trừ phần bị đếm trùng do ký tự này đã từng xuất hiện trước đó.
        if seen[c]:
            new = (new - last[c]) % MOD
        last[c] = dp
        seen[c] = True
        dp = new

    # Trừ 1 để loại chuỗi rỗng.
    ans = (dp - 1) % MOD
    sys.stdout.write(str(ans) + "\n")


main()

File: cses/08_string_algorithms/1149_distinct-subsequences/solution.cpp
#include <array>
#include <iostream>
#include <string>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    constexpr long long MOD = 1'000'000'007LL;
    std::string s;
    std::cin >> s;

    // dp: số subsequence phân biệt tính cả chuỗi rỗng; khởi tạo chỉ có chuỗi rỗng.
    long long subsequences = 1;
    // previous_count[c] = giá trị dp ngay trước lần xuất hiện gần nhất của c (dp[j-1]).
    std::array<long long, 26> previous_count{};
    std::array<bool, 26> seen{};

    for (const char character : s) {
        const std::size_t letter = static_cast<std::size_t>(character - 'a');
        // Gấp đôi: mỗi chuỗi cũ có thể giữ nguyên hoặc gắn thêm ký tự hiện tại.
        long long next = 2 * subsequences % MOD;
        // Trừ phần bị đếm trùng do ký tự này đã từng xuất hiện trước đó.
        if (seen[letter]) {
            next = (next - previous_count[letter] + MOD) % MOD;
        }
        previous_count[letter] = subsequences;
        seen[letter] = true;
        subsequences = next;
    }

    // Trừ 1 để loại chuỗi rỗng, cộng MOD giữ kết quả không âm.
    std::cout << (subsequences - 1 + MOD) % MOD << '\n';
    return 0;
}
